### ISSUE TITLE
Server stops responding after several minutes on Windows 10

### ISSUE THREAD

**Issue by randy909** [eclipse/jetty.project#3613]
1) Run the ManyConnectors example here: https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/examples/embedded/src/main/java/org/eclipse/jetty/embedded/ManyConnectors.java
2) Wait about 5 minutes.
3) Try connecting with any tool and the connection will hang forever. It will work normally if you have not waited long enough. Both http and https connections will hang.

Windows 10 version 1709 build 16299.1029, Java 8u201, jetty 9.4.18.v20190429.

**Downgrading jetty to 9.3.27.v20190418 fixes the issue.**

We captured some logs. Here is a good request:
```2019-04-30 11:42:01.342:DBUG:oeji.ManagedSelector:qtp29463805-44: Selector sun.nio.ch.WindowsSelectorImpl@19b05d9 woken up from select, 1/1/1 selected
2019-04-30 11:42:01.342:DBUG:oeji.ManagedSelector:qtp29463805-44: Selector sun.nio.ch.WindowsSelectorImpl@19b05d9 processing 1 keys, 0 updates
2019-04-30 11:42:01.342:DBUG:oeji.ManagedSelector:qtp29463805-44: selected 1 sun.nio.ch.SelectionKeyImpl@15ced5a SocketChannelEndPoint@d1a45e{/127.0.0.1:61335<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=16301/30000}{io=1/1,kio=1,kro=1}->HttpConnection@199f898[p=HttpParser{s=START,0 of -1},g=HttpGenerator@db5021{s=START}]=>HttpChannelOverHttp@ee271c{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0} 
2019-04-30 11:42:01.342:DBUG:oeji.ChannelEndPoint:qtp29463805-44: onSelected 1->0 r=true w=false for SocketChannelEndPoint@d1a45e{/127.0.0.1:61335<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=16315/30000}{io=1/0,kio=1,kro=1}->HttpConnection@199f898[p=HttpParser{s=START,0 of -1},g=HttpGenerator@db5021{s=START}]=>HttpChannelOverHttp@ee271c{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0}
2019-04-30 11:42:01.356:DBUG:oeji.ChannelEndPoint:qtp29463805-44: task CEP:SocketChannelEndPoint@d1a45e{/127.0.0.1:61335<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=16315/30000}{io=1/0,kio=1,kro=1}->HttpConnection@199f898[p=HttpParser{s=START,0 of -1},g=HttpGenerator@db5021{s=START}]=>HttpChannelOverHttp@ee271c{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0}:runFillable:BLOCKING
2019-04-30 11:42:01.356:DBUG:oejut.ReservedThreadExecutor:qtp29463805-44: ReservedThreadExecutor@b5d3e{s=3/8,p=0} tryExecute EatWhatYouKill@1f1220/SelectorProducer@23446d/PRODUCING/p=false/QueuedThreadPool[qtp29463805]@1c194fd{STARTED,8<=15<=200,i=2,q=0}[ReservedThreadExecutor@b5d3e{s=3/8,p=0}][pc=0,pic=0,pec=0,epc=1]@2019-04-30T11:42:01.356-05:00
2019-04-30 11:42:01.356:DBUG:oejut.ReservedThreadExecutor:qtp29463805-44: ReservedThreadExecutor@b5d3e{s=2/8,p=0}@18def68 offer EatWhatYouKill@1f1220/SelectorProducer@23446d/PRODUCING/p=false/QueuedThreadPool[qtp29463805]@1c194fd{STARTED,8<=15<=200,i=2,q=0}[ReservedThreadExecutor@b5d3e{s=2/8,p=0}][pc=0,pic=0,pec=0,epc=1]@2019-04-30T11:42:01.356-05:00
2019-04-30 11:42:01.356:DBUG:oejuts.EatWhatYouKill:qtp29463805-44: EatWhatYouKill@1f1220/SelectorProducer@23446d/IDLE/p=true/QueuedThreadPool[qtp29463805]@1c194fd{STARTED,8<=15<=200,i=2,q=0}[ReservedThreadExecutor@b5d3e{s=2/8,p=0}][pc=0,pic=0,pec=0,epc=1]@2019-04-30T11:42:01.356-05:00 m=EXECUTE_PRODUCE_CONSUME t=CEP:SocketChannelEndPoint@d1a45e{/127.0.0.1:61335<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=16315/30000}{io=1/0,kio=1,kro=1}->HttpConnection@199f898[p=HttpParser{s=START,0 of -1},g=HttpGenerator@db5021{s=START}]=>HttpChannelOverHttp@ee271c{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0}:runFillable:BLOCKING/BLOCKING
2019-04-30 11:42:01.356:DBUG:oejut.ReservedThreadExecutor:qtp29463805-88: ReservedThreadExecutor@b5d3e{s=2/8,p=0}@18def68 task=EatWhatYouKill@1f1220/SelectorProducer@23446d/IDLE/p=true/QueuedThreadPool[qtp29463805]@1c194fd{STARTED,8<=15<=200,i=2,q=0}[ReservedThreadExecutor@b5d3e{s=2/8,p=0}][pc=0,pic=0,pec=0,epc=1]@2019-04-30T11:42:01.356-05:00
2019-04-30 11:42:01.356:DBUG:oeji.FillInterest:qtp29463805-44: fillable FillInterest@1445861{AC.ReadCB@199f898{HttpConnection@199f898::SocketChannelEndPoint@d1a45e{/127.0.0.1:61335<->/127.0.0.1:8080,OPEN,fill=FI,flush=-,to=16315/30000}{io=1/0,kio=1,kro=1}->HttpConnection@199f898[p=HttpParser{s=START,0 of -1},g=HttpGenerator@db5021{s=START}]=>HttpChannelOverHttp@ee271c{r=1,c=false,c=false/false,a=IDLE,uri=null,age=0}}}
2019-04-30 11:42:01.356:DBUG:oejuts.EatWhatYouKill:qtp29463805-88: EatWhatYouKill@1f1220/SelectorProducer@23446d/IDLE/p=true/QueuedThreadPool[qtp29463805]@1c194fd{STARTED,8<=15<=200,i=2,q=0}[ReservedThreadExecutor@b5d3e{s=2/8,p=0}][pc=0,pic=0,pec=0,epc=2]@2019-04-30T11:42:01.356-05:00 tryProduce true
2019-04-30 11:42:01.356:DBUG:oeji.ManagedSelector:qtp29463805-88: updateable 0
2019-04-30 11:42:01.356:DBUG:oeji.ManagedSelector:qtp29463805-88: updates 0
...
```

And a hung request
```2019-04-30 11:55:17.984:DBUG:oeji.ManagedSelector:qtp29463805-51-acceptor-0@1146f8a-ServerConnector@117ee06{SSL,[ssl, http/1.1]}{0.0.0.0:8443}: Queued change org.eclipse.jetty.io.ManagedSelector$Accept@d30acb on ManagedSelector@17b85fc{STARTED} id=3 keys=0 selected=0 updates=0
2019-04-30 11:55:17.984:DBUG:oeji.ManagedSelector:qtp29463805-51-acceptor-0@1146f8a-ServerConnector@117ee06{SSL,[ssl, http/1.1]}{0.0.0.0:8443}: Wakeup on submit ManagedSelector@17b85fc{STARTED} id=3 keys=0 selected=0 updates=1
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: Selector sun.nio.ch.WindowsSelectorImpl@156375b woken with none selected
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: Selector sun.nio.ch.WindowsSelectorImpl@156375b woken up from select, 0/0/0 selected
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: Selector sun.nio.ch.WindowsSelectorImpl@156375b processing 0 keys, 1 updates
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: updateable 1
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: update org.eclipse.jetty.io.ManagedSelector$Accept@d30acb
2019-04-30 11:55:17.993:DBUG:oejut.QueuedThreadPool:qtp29463805-50: queue org.eclipse.jetty.io.ManagedSelector$Accept@d30acb
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: updates 0
2019-04-30 11:55:17.993:DBUG:oeji.ManagedSelector:qtp29463805-50: Selector sun.nio.ch.WindowsSelectorImpl@156375b waiting with 1 keys
```

So no mention of a `SocketChannelEndPoint` or `ReservedThreadExecutor` in the bad example. Seems the accept code is still running but there are no threads to hand the new sockets off to?

Works just fine on my Mac.

**Comment by joakime** [comment]
```
Selector sun.nio.ch.WindowsSelectorImpl@156375b woken up from select, 0/0/0 selected
```

Seems like a duplicate of the JVM bug reported in issue #2205

**Comment by sbordet** [comment]
@joakime from the little that has been reported it's not a duplicate of #2205 because the selector has been woken up due to 1 update to process.
@joakime can you reproduce on a Windows box?
@randy909 can you attach the full DEBUG logs?

**Comment by joakime** [comment]
I'm actively attempting to replicate on a box running Windows 10 Version 1803 (OS Build 17134.706) - with Oracle JDK 8u202

**Comment by joakime** [comment]
I'm unable to replicate.
Using Windows 10 Version 1803 (OS Build 17134.706)
With Oracle JDK 8u202.

I started `ManyConnectors` from :
https://github.com/eclipse/jetty.project/blob/jetty-9.4.18.v20190429/examples/embedded/src/main/java/org/eclipse/jetty/embedded/ManyConnectors.java

I waited 10 minutes (timed with kitchen timer) and then made a request to both port 8080 and 8443 without issue.

Logs: https://gist.github.com/joakime/ec0e150ebd88d1765731c7813a303e29

**Comment by randy909** [comment]
I missed a few details:
1) You must modify ManyConnectors.java to use DefaultHandler instead of HelloHandler.
2) You must launch from an eclipse project built by eclipse (i.e. not mvn or gradle)

I captured some logs, one showing the eclipse native build (hangs) and one showing the eclipse gradle-plugin build (works).

I can push an eclipse project that fails if that helps.

Both logs: https://gist.github.com/randy909/b7e74a15ccca3193b959b838a09cb78d

**Comment by joakime** [comment]
Still unable to replicate.

Current process ...

1. Using Windows 10 - Version 1803 (OS Build 17134.706)
2. Using Java from Oracle JDK 8u202
3. Opened all of Eclipse Jetty from Eclipse IDE 2019-03
4. Switched to tag `jetty-9.4.18.v20190429` (this is the tag for the release)
5. Enabled DEBUG logging on `examples/embedded`
6. Replace `HelloHandler` with `DefaultHandler` in `ManyConnectors.java`
7. Start `ManyConnectors` from IDE
8. Wait 10 minutes
9. Hit both `http://localhost:8080` and `https://localhost:8443`

No hang, it still responds.
Using Chrome 74.0.3729.131 (Official Build) (64-bit) or Firefox 66.0.3 (64-bit) works.
Using curl or wget from other machine over network works too.
Using Microsoft Edge 42.17134.1.0 also works.
Using ab (apache bench) against this server works splendidly as well, still no hang or unresponsiveness.

At this point I suspect that you have some firewall or other extra component on your windows machine that is causing your behavior.

**Comment by randy909** [comment]
Thank you for all your efforts to replicate! It is most definitely not a firewall issue on my end: I've used wireshark to analyze the traffic and the HTTP GET is making it to the server, the firewall is disabled for private networks and I'm hitting localhost, and I'm able to reliably make it fail/pass by running the test the ways I've described (although not in enough detail I think and I'll try to fix that below). Also worth mentioning is that we've seen this happen on several machines in our build cluster. Our app is an RCP app in which we embed a jetty server.

I just noticed that I am one version back on my failing project, 9.4.17.v20190418, but I'm at 9.4.18 on the working, gradle-based one. Is it possible that this was fixed in the latest release? I will try again tomorrow.

Possibilities from my perspective:
1) Not reproducible on your newer version of Windows
2) Not reproducible on your jdk8_202 vs my jdk8_201
3) Not reproducible without the same classpath as mine (yours has many more libs)
4) It was fixed in the latest patch

If the latest patch doesn't work I will push a project for you to try. I'll also try jdk8 202.

It's probably worth mentioning that I've tried *many* different configurations and a lot of them do work well so it doesn't surprise me that it's hard for you to reproduce it. This is a finicky one! :)

**Comment by joakime** [comment]
Lets use https://github.com/jetty-project/embedded-jetty-cookbook to coordinate.
That's a smaller project, and has the `ManyConnectors` example project as well.

Same test as before, with that project's `ManyConnectors` also shows no issues for me on Windows 10.

**Comment by randy909** [comment]
9.4.18 fixed it, I was using 9.4.17 on the failing project. Sorry for the confusion. We downloaded the latest Jetty from the eclipse website and it showed the latest as 9.4.17 on Tuesday whereas maven central had 9.4.18. Disregard the other variables I mentioned like eclipse, gradle, jdk version, etc. I was able to reproduce the failure on all my configurations with 9.4.17 and fix it with 9.4.18.

My guess is that it was fixed on #3550.

Thank you for all your help!
